### PR TITLE
Add Chromatogram functionality

### PR DESCRIPTION
**Pull request by jotsetung**
- Add `Chromatogram` and `Chromatograms` classes.
- Add `chromatogram,MSnExp` method to extract a `Chromatograms` object from an `MSnExp` (or `OnDiskMSnExp`.
- Add `plot,Chromatogram` and `plot,Chromatograms` methods.
- Add `filterRt,Chromatogram` method.
- Add `clean,Chromatogram` method.

### CODE REVIEW

File: R/functions-MSnExp.R [hunk 1]
```diff
@@ -351,3 +351,28 @@ removeReporters_MSnExp <- function(object, reporters = NULL,
...
+    ## Replicate if nrow rt is 1 to match nrow of mz.
+    if (nrow(rt) == 1)
+        rt <- matrix(rep(rt, nrow(mz)), ncol = 2, byrow = TRUE)
+    if (ncol(mz) != 2)
+        mz <- t(apply(mz, MARGIN = 1, range))
+    if (nrow(mz) == 1)
+        mz <- matrix(rep(mz, nrow(rt)), ncol = 2, byrow = TRUE)
+    if (nrow(rt) != nrow(mz))
+        stop("dimensions of 'rt' and 'mz' have to match")
+    ## Check mz
```
Comment (sgibb) [inline R/functions-MSnExp.R]: Is here something missing? Is this function supposed to return something? Is it used/needed at all?

File: R/functions-MSnExp.R [hunk 2]
```diff
@@ -351,3 +351,28 @@ removeReporters_MSnExp <- function(object, reporters = NULL,
     return(object)
 }
 
+.chromatogram <- function(x, rt, mz, aggregationFun = "sum",
+                          missingValue = NA_real_) {
+    missingValue <- as.numeric(missingValue)
```
Comment (sgibb) [inline R/functions-MSnExp.R]: Why is `aggregationFun` and `missingValue` needed? They are not used.

File: R/functions-OnDiskMSnExp.R [hunk 3]
```diff
@@ -481,3 +481,201 @@ precursorValue_OnDiskMSnExp <- function(object, column) {
...
+    if (nrow(rt) == 1)
+        rt <- matrix(rep(rt, nrow(mz)), ncol = 2, byrow = TRUE)
+    if (ncol(mz) != 2)
+        mz <- t(apply(mz, MARGIN = 1, range))
+    if (nrow(mz) == 1)
+        mz <- matrix(rep(mz, nrow(rt)), ncol = 2, byrow = TRUE)
+    if (nrow(rt) != nrow(mz))
+        stop("dimensions of 'rt' and 'mz' have to match")
+    ## Identify indices of all spectra that are within the rt ranges.
+    rtimes <- rtime(x)
```
Comment (sgibb) [inline R/functions-OnDiskMSnExp.R]: This looks similar to `.chromatogram`. Is `.chromatogram` used/needed at all?

File: R/functions-OnDiskMSnExp.R [hunk 4]
```diff
@@ -481,3 +481,201 @@ precursorValue_OnDiskMSnExp <- function(object, column) {
...
+
+    ## 1) Subset x keeping all spectra that fall into any of the provided rt
+    ##    ranges.
+    keep_idx <- unlist(apply(rt, MARGIN = 1, function(z)
+        which(rtimes >= z[1] & rtimes <= z[2])), use.names = FALSE)
```
Comment (sgibb) [inline R/functions-OnDiskMSnExp.R]: I assume `rtimes` is sorted. Is there a guarantee that `rt` is sorted (meaning `rt[, 1] < rt[,2]`?):
If that is true we could reduce the comparisons and loops runs by removing overlapping `rt` ranges before (here and below in `bpmapply`). Depending on `nrow(rt)` this could be faster. Not sure if needed, e.g.:
```r
rt <- matrix(c(50, 100, 300, 400, 350, 500, 700, 900), ncol=2, byrow=TRUE)
#      [,1] [,2]
# [1,]   50  100
# [2,]  300  400
# [3,]  350  500
# [4,]  700  900
overlap <- rt[-1 ,1] <= rt[-nrow(rt), 2]
# [1] FALSE  TRUE FALSE
cbind(rt[c(TRUE, !overlap), 1L], rt[c(!overlap, TRUE), 2L])
#      [,1] [,2]
# [1,]   50  100
# [2,]  300  500
# [3,]  700  900
```
(Maybe as a new `utils....` function with some more test cases (at least some edge cases)?)

File: R/methods-Chromatograms.R [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                         j <- match(j, colnames(x))
+                     ## check i and j - currently we support only replacing
+                     ## single elements, rows or columns.
+                     if (length(i) > 1 & length(j) > 1)
+                         stop("Only replacement of single elements, rows or ",
+                              "columns supported")
+                     ## Check value.
+                     if (is(value, "Chromatogram"))
+                         value <- list(value)
+                     if (!all(sapply(value, function(z) is(z, "Chromatogram"))))
```
Comment (sgibb) [inline R/methods-Chromatograms.R]: I would use `res <- vapply(x, is, logical(1L), "Chromatogram")` (or if you don't like `vapply`: `res <- sapply(x, is, "Chromatogram")`.

File: R/functions-OnDiskMSnExp.R (hunk 4 above)
Comment (sgibb) [inline R/functions-OnDiskMSnExp.R]: Stupid me. That's already available in `IRanges::reduce`:
```r
ir <- IRanges(rt[,1], rt[,2])
# IRanges object with 4 ranges and 0 metadata columns:
#           start       end     width
#       <integer> <integer> <integer>
#   [1]        50       100        51
#   [2]       300       400       101
#   [3]       350       500       151
#   [4]       700       900       201
reduce(ir)
# IRanges object with 3 ranges and 0 metadata columns:
#           start       end     width
#       <integer> <integer> <integer>
#   [1]        50       100        51
#   [2]       300       500       201
#   [3]       700       900       201
```

So I vote to call `IRanges::reduce` on `rt` if it is sorted as mentioned above.

Comment (sgibb) [inline R/functions-OnDiskMSnExp.R]: The whole `apply` could IMHO be replaced by `queryHits(findOverlaps(rtimes, reduce(ir)))`

**Comment by jotsetung** [comment]
Thanks @sgibb for the great review (as always 👍).

File: R/functions-Chromatograms.R [hunk 6]
```diff
@@ -0,0 +1,104 @@
+## functions for Chromatograms class
+
+.validChromatograms <- function(x) {
+    msg <- character()
```
Comment (jotsetung) [inline R/functions-Chromatograms.R]: adopted from Martin Morgan's *new* coding style - I'm trying to use wherever possible base functions. I'm however also fine with changing that to `Biobase::validMsg`.

File: R/functions-Chromatograms.R [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    ## All elements have to be of type Chromatogram
+    if (length(x)) {
+        res <- sapply(x, FUN = function(z) is(z, "Chromatogram"))
```
Comment (jotsetung) [inline R/functions-Chromatograms.R]: 👍

File: R/functions-Chromatograms.R [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        if (!all(res))
+            msg <- c(msg, paste0("All elements have to be of type ",
+                                 "'Chromatogram'."))
+        ## Shall we also ensure that fromFile in each column is the same?
+    }
+    if (length(msg))
+        msg
+    else TRUE
```
Comment (jotsetung) [inline R/functions-Chromatograms.R]: 👍

File: R/functions-Chromatograms.R [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    if (is.null(main)) {
+        suppressWarnings(
+            mzr <- range(lapply(x, mz), na.rm = TRUE, finite = TRUE)
+        )
+        main <- paste0(format(mzr, digits = 7), collapse = " - ")
+    }
+    ## Number of measurements we've got per chromatogram. This can be different
+    ## between samples, from none (if not a single measurement in the rt/mz)
+    ## to the number of data points that were actually measured.
+    lens <- unique(lengths(x))
```
Comment (jotsetung) [inline R/functions-Chromatograms.R]: 👍

File: R/functions-Chromatograms.R (hunk 6 above)
Comment (sgibb) [inline R/functions-Chromatograms.R]: Ok, I was not aware of a new coding style (do you have a link to Martin Morgan's coding style?). But I really like the idea to use as much `base` functions as possible. So go ahead. Let's forget `Biobase::validMsg`.

Comment (jotsetung) [inline R/functions-Chromatograms.R]: I had the `Biobase::validMsg` calls in the validation function of the classes in https://github.com/Bioconductor/AnnotationFilter , but Martin removed them all :-)

File: R/functions-Chromatograms.R [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    ## Same for the intensities.
+    ints <- do.call(cbind, lapply(x, function(z) {
+        cur_len <- length(z)
+        if (length(z) == 0)
+            max_len_vec
+        else {
+            max_len_vec[seq_len(cur_len)] <- intensity(z)
+            max_len_vec
+        }
+    }))
```
Comment (jotsetung) [inline R/functions-Chromatograms.R]: 👍

File: R/functions-Chromatogram.R [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+#' @return \code{TRUE} if the \code{object} is valid and the error messages
+#'     otherwise (i.e. a \code{character}).
+#' 
+#' @author Johannes Rainer
+#' 
+#' @noRd
+.validChromatogram <- function(object) {
+    msg <- character()
+    if (length(object@rtime) != length(object@intensity))
+        msg <- c(msg, "Length of 'rt' and 'intensity' have to match!")
```
Comment (jotsetung) [inline R/functions-Chromatogram.R]: See above

File: R/functions-MSnExp.R (hunk 2 above)
Comment (jotsetung) [inline R/functions-MSnExp.R]: Sorry, I first thought of implementing this function for `MSnExp` objects but then decided to go for one function that can be used for both `MSnExp` and `OnDiskMSnExp` (i.e. the `extractMultipleChromatograms` function. Just forgot to remove this function.

File: R/functions-OnDiskMSnExp.R (hunk 4 above)
Comment (jotsetung) [inline R/functions-OnDiskMSnExp.R]: `rtimes` is not sorted, it is sorted within each file, but the `OnDiskMSnExp`/`MSnExp` object is supposed to contain spectra from several files. So it will be retention times from first file, then retention times from second file and so on.

Regarding the solution using `IRanges`: I principally like the idea, but creating the `IRanges` objects (since they are `S4` might not be the fastest), also, I am not sure if it would work, as `IRanges` works AFAIK only with integer ranges - both the `rt` and `rtimes` are however `numeric`.

Your suggestion is however very welcome - I will investigate if I can improve the code.

File: R/methods-Chromatograms.R (hunk 5 above)
Comment (jotsetung) [inline R/methods-Chromatograms.R]: 👍

File: R/functions-Chromatograms.R (hunk 10 above)
Comment (jotsetung) [inline R/functions-Chromatograms.R]: This definitely simplified the code, thanks @sgibb . Only thing is I'll still plot an empty plot instead of stopping with an error if no data is available to plot. I somehow don't like `plot` calls throwing an error.

File: R/functions-OnDiskMSnExp.R (hunk 4 above)
Comment (jotsetung) [inline R/functions-OnDiskMSnExp.R]: Unfortunately there was not much I could do. I simplified the function but the `apply` and `for` loop can not (easily) be replaced.

**Comment by sgibb** [comment]
@jotsetung great!

**Comment by jotsetung** [comment]
Let's wait from the final approval from @lgatto

**Comment by lgatto** [comment]
I have been teaching and giving a talk end of the week. Will look at it asap.

**Comment by lgatto** [comment]
You added the classes in the devel vignette. Wouldn't it be useful to also mention them in either the main demo vignette, or have a new one?

**Comment by jotsetung** [comment]
I'll add it to the main vignette - but now I'm going on holidays, so it will have to wait... hope I don't forget it in the meantime

**Comment by lgatto** [comment]
No problem re vignette - I have created an issue #224 that I assigned to you ;-)

**Comment by lgatto** [comment]
Thank you very much @jotsetung for this!

File: R/AllGenerics.R [hunk 12]
```diff
@@ -143,3 +143,9 @@ setGeneric("filterEmptySpectra", function (object, ...) standardGeneric("filterE
 setGeneric("bpi", function(object, ...) standardGeneric("bpi"))
 setGeneric("spectrapply", function(object, ...) standardGeneric("spectrapply"))
 setGeneric("splitByFile", function(object, f, ...) standardGeneric("splitByFile"))
+
+## Chromatogram class:
+setGeneric("productMz", function(object, value) standardGeneric("productMz"))
+setGeneric("aggregationFun", function(object, ...)
```
Comment (jotsetung) [inline R/AllGenerics.R]: Done in https://github.com/lgatto/MSnbase/commit/978e686f6920061d8c705220cdd6eb21729a3882
